playwright/src/main/java/com/microsoft/playwright/junit/impl/PlaywrightExtension.java

```diff
@@ -4,19 +4,51 @@
 import com.microsoft.playwright.junit.Options;
 import org.junit.jupiter.api.extension.*;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
 import static com.microsoft.playwright.junit.impl.ExtensionUtils.isParameterSupported;
 import static com.microsoft.playwright.junit.impl.ExtensionUtils.setTestIdAttribute;
 
-public class PlaywrightExtension implements ParameterResolver, AfterAllCallback {
-  private static final ThreadLocal<Playwright> threadLocalPlaywright = new ThreadLocal<>();
+public class PlaywrightExtension implements ParameterResolver, BeforeAllCallback, ExtensionContext.Store.CloseableResource {
+  private static final ThreadLocal<Playwright> threadLocalPlaywright;
+  private static final List<Playwright> playwrightList;
+  private final static Lock beforeLock;
+  private static boolean isTestRunStarted;
+
+  static {
+    isTestRunStarted = false;
+    beforeLock = new ReentrantLock();
+    threadLocalPlaywright = new ThreadLocal<>();
+    playwrightList = Collections.synchronizedList(new ArrayList<>());
+  }
 
+  // Before a Test class starts, we register a closeable resource
+  // We use a lock + boolean to ensure this only gets run once for the entire test run
   @Override
-  public void afterAll(ExtensionContext extensionContext) {
-    Playwright playwright = threadLocalPlaywright.get();
-    threadLocalPlaywright.remove();
-    if (playwright != null) {
+  public void beforeAll(final ExtensionContext context) {
+    try {
+      beforeLock.lock();
```

Comment:
Correct me if I am wrong, but as far as I understand we don't need neither beforeLock nor isTestRunStarted here because PlaywrightExtension instance is created once per test class and for that test class `beforeAll` is also called exactly once. `ExtensionContext.Store` should already handle concurrent access as it can be access by multiple tests. Their [source code](https://github.com/sbrannen/junit5/blob/84ac590dd9bdee656bfbfd45aa7afbdfd26a3478/junit-platform-engine/src/main/java/org/junit/platform/engine/support/store/NamespacedHierarchicalStore.java#L49
) matches my understanding.

Reply:
You are correct that beforeAll is called once per test class, however we want this particular beforeAll to be called once per test run and register the closeable resource once.

The lock is to ensure if two test classes hit the beforeAll at the same time, only one of them can register the closeable resource.

If we don't do it this way, every test class will register its own closeable resource and the close method will be called after each test class completes, which is essentially an AfterAll, and that's what we don't want to do.

We want to close all playwright instances after the entire test run is complete.

Reply:
I see what you mean. In that case how about we introduce a class that manages the lifetime of the Playwright objects and make it closable:

```java
  class PlaywrgihtInstances implements ExtensionContext.Store.CloseableResource {
    private List<Playwright> playwrightList;

    static synchronized PlaywrgihtInstances getOrCreateFor(ExtensionContext.Store rootStore) {
      if (rootStore.get(PlaywrgihtInstances.class) != null) {
        rootStore.put(PlaywrgihtInstances.class, new PlaywrgihtInstances());
      }
    }

    @Override
    public void close() throws Throwable {
      ...
    }
  }
```

The you just need to create playwrights via this class to make sure all of them are registered right after creation. WDYT?

Reply:
i feel like this would make it less performant.  getOrCreate can be called many different times over a test class.  if a test class 5 methods, it could be called 5 times.  where as the beforeAll is only called once per test class and never more.

maybe i'm not seeing the benefit of this.  would you mind implementing what you are suggesting in this branch.  following the code would make it easier for me to understand.

Reply:
It'd only need to be called where `playwrightList.add(` is currently called. Basically the same number of times as the number of playwright instances created. I can send a follow-up PR with the change, no problem.